### PR TITLE
Correct IScriptsManager resolution in Settings dialog

### LINKED ISSUE #11253: FormSettings: InvalidOperationException IGitUICommands
In the current `master`, whenever I open the settings window, I get an exception here:
https://github.com/gitextensions/gitextensions/blob/a0d8a30c79ba54e5b231786f46d381b96c584955/GitUI/CommandsDialogs/SettingsDialog/Pages/ScriptsSettingsPage.cs#L124

I can push the rebased version anyway, but I feel iffy about pushing a version that I can only test partially.

_Originally posted by @SlugFiller in https://github.com/gitextensions/gitextensions/issues/11239#issuecomment-1751968802_

### PR DESCRIPTION
**Pull request by RussKie**
Resolves #11253

As reported in https://github.com/gitextensions/gitextensions/pull/11239#issuecomment-1751968802 (/cc: @SlugFiller)

### CODE REVIEW

File: GitUI/GitModuleControl.cs [hunk 1]
```diff
@@ -82,20 +82,6 @@ public IGitUICommandsSource UICommandsSource
...
-        /// <para>This method will not attempt to initialise the commands if they have not
-        /// yet been initialised.</para>
-        /// <para>By contrast, the <see cref="UICommands"/> property attempts to initialise
-        /// the value if not previously initialised.</para>
-        /// </remarks>
-        public bool TryGetUICommands([NotNullWhen(returnValue: true)] out GitUICommands? commands)
-        {
-            commands = _uiCommandsSource?.UICommands;
-            return commands is not null;
-        }
```
Comment (mstv) [inline GitUI/GitModuleControl.cs:85]: StyleCop:
```suggestion
```

**Review by mstv — COMMENTED** [review]
works for me

BTW: The ScriptsSettingsPage takes much time for loading.

**Comment by RussKie** [comment]
I think, it's always been quite slow.

There could be few opportunities to improve by making the scripts loading
async, as well as moving away from xml serialisation.

**Comment by SlugFiller** [comment]
Well, the settings page opens now. But the script settings aren't being saved. Going to try to investigate.

**Comment by SlugFiller** [comment]
Okay, I figured it out. `GitUI.ScriptsEngine.IScriptsManager` does not offer an interface for altering the scripts. Therefore, in `GitUI.CommandsDialogs.SettingsDialog.Pages.ScriptsSettingsPage.PageToSettings` it creates a scripts list that is never used.

Also, I hate everything about the way `GitUI.ScriptsEngine.GetScripts.FixAmbiguousHotkeyCommandIdentifiers` is implemented. Instead of using the `HashSet` to find empty ids (in a defragmenting fashion), it does a max operation on the original set. Even if only doing max, it doesn't try to simply store the max and increment it.

Last, but not least, it should probably not be inside a function, so it can be used to do a sanity check on saved settings, and not just loaded ones.

Finally, the having a proxy version of the script binding interface should have probably been reserved for being able to deserialize older versions of the XML (e.g. to rename options, or change their type).

**Comment by SlugFiller** [comment]
Should I open a PR about script saving, or you have it covered? With minimal changes (not amazing, but works), I have this:
```diff
diff --git a/GitUI/CommandsDialogs/SettingsDialog/Pages/ScriptsSettingsPage.cs b/GitUI/CommandsDialogs/SettingsDialog/Pages/ScriptsSettingsPage.cs
index 84ba78067..70545cf34 100644
--- a/GitUI/CommandsDialogs/SettingsDialog/Pages/ScriptsSettingsPage.cs
+++ b/GitUI/CommandsDialogs/SettingsDialog/Pages/ScriptsSettingsPage.cs
@@ -198,6 +198,8 @@ protected override void PageToSettings()
                 scripts.Add(proxy);
             }
 
+            _scriptsManager.SetScripts(scripts);
+
             AppSettings.OwnScripts = _scriptsManager.SerializeIntoXml();
 
             base.PageToSettings();
diff --git a/GitUI/ScriptsEngine/IScriptsManager.cs b/GitUI/ScriptsEngine/IScriptsManager.cs
index 18288b17e..d26ac208d 100644
--- a/GitUI/ScriptsEngine/IScriptsManager.cs
+++ b/GitUI/ScriptsEngine/IScriptsManager.cs
@@ -8,6 +8,8 @@ public interface IScriptsManager
 
         BindingList<ScriptInfo> GetScripts();
 
+        void SetScripts(List<ScriptInfo> scripts);
+
         string SerializeIntoXml();
     }
 }
diff --git a/GitUI/ScriptsEngine/ScriptsManager.cs b/GitUI/ScriptsEngine/ScriptsManager.cs
index 483f34889..3f581368d 100644
--- a/GitUI/ScriptsEngine/ScriptsManager.cs
+++ b/GitUI/ScriptsEngine/ScriptsManager.cs
@@ -39,6 +39,16 @@ static void FixAmbiguousHotkeyCommandIdentifiers(BindingList<ScriptInfo> loadedS
             }
         }
 
+        public void SetScripts(List<ScriptInfo> scripts)
+        {
+            _scripts.Clear();
+
+            foreach (var script in scripts)
+            {
+                _scripts.Add(script);
+            }
+        }
+
         public ScriptInfo? GetScript(int scriptId)
         {
             foreach (ScriptInfo script in GetScripts())
```

File: GitUI/CommandsDialogs/SettingsDialog/Pages/ScriptsSettingsPage.cs [hunk 2]
```diff
@@ -191,15 +182,15 @@ protected override void PageToSettings()
         {
             // TODO: this is an abomination, the whole script persistence must be scorched and rewritten
 
-            List<ScriptInfo> scripts = new(ScriptsManager.GetScripts());
+            BindingList<ScriptInfo> scripts = _scriptsManager.GetScripts();
```
Comment (RussKie) [inline GitUI/CommandsDialogs/SettingsDialog/Pages/ScriptsSettingsPage.cs:185]: This was the cause of the regression.

**Comment by SlugFiller** [comment]
Confirmed saving settings works now

**Comment by mstv** [comment]
saving works but many tests fail

**Comment by RussKie** [comment]
/fml
I'll have a look tonight

**Comment by RussKie** [comment]
The PR build succeeded though the master is now failing.... I will have to have a look tomorrow or the day after.

**Pull request merged**